browser/net/brave_proxying_url_loader_factory.cc

```diff
@@ -0,0 +1,969 @@
+    // intuitive), |onBeforeSendHeaders| is only dispatched for HTTP and HTTPS
+    // requests.
+    // TODO(iefremov): Bind real variables and modify the corresponding API of
+    // callbacks. For now it is safe to leave it as is, since these headers are
+    // only needed in non |header_client_binding_| codepath, but currently
+    // callbacks do meaningful work only for real requests that should go
+    // though header client.
+    auto continuation = base::BindRepeating(
+        &InProgressRequest::ContinueToSendHeaders, weak_factory_.GetWeakPtr(),
+        std::set<std::string>(), std::set<std::string>());
+
+    ctx_ = std::make_shared<brave::BraveRequestInfo>();
+    brave::BraveRequestInfo::FillCTX(request_, render_process_id_,
+                                     frame_tree_node_id_, request_id_,
+                                     resource_context_, ctx_);
+    int result = factory_->request_handler_->OnBeforeStartTransaction(
+        ctx_, continuation, &request_.headers);
+
+    if (result == net::ERR_BLOCKED_BY_CLIENT) {
+      // The request was cancelled synchronously. Dispatch an error notification
+      // and terminate the request.
+      OnRequestError(network::URLLoaderCompletionStatus(result));
+      return;
+    }
+
+    if (result == net::ERR_IO_PENDING) {
+      // One or more listeners is blocking, so the request must be paused until
+      // they respond. |continuation| above will be invoked asynchronously to
+      // continue or cancel the request.
+      //
+      // We pause the binding here to prevent further client message processing.
+      if (proxied_client_binding_.is_bound())
+        proxied_client_binding_.PauseIncomingMethodCallProcessing();
+      return;
+    }
+    DCHECK_EQ(net::OK, result);
+  }
+
+  ContinueToSendHeaders(std::set<std::string>(), std::set<std::string>(),
+                        net::OK);
+}
+
+void BraveProxyingURLLoaderFactory::InProgressRequest::ContinueToStartRequest(
+    int error_code) {
+  if (error_code != net::OK) {
+    OnRequestError(network::URLLoaderCompletionStatus(error_code));
+    return;
+  }
+
+  if (current_request_uses_header_client_ && !redirect_url_.is_empty()) {
+    HandleBeforeRequestRedirect();
+    return;
+  }
+
+  if (proxied_client_binding_.is_bound())
+    proxied_client_binding_.ResumeIncomingMethodCallProcessing();
+
+  if (header_client_binding_)
+    header_client_binding_.ResumeIncomingMethodCallProcessing();
+
+  if (!target_loader_.is_bound() && factory_->target_factory_.is_bound()) {
+    // Nothing has cancelled us up to this point, so it's now OK to
+    // initiate the real network request.
+    network::mojom::URLLoaderClientPtr proxied_client;
+    proxied_client_binding_.Bind(mojo::MakeRequest(&proxied_client));
+    uint32_t options = options_;
+    // Even if this request does not use the header client, future redirects
+    // might, so we need to set the option on the loader.
+    if (has_any_extra_headers_listeners_)
```

Comment:
per the comment below, I think this should just be a method that checks the request url and selectively adds this. `has_any_extra_headers_listeners_` doesn't really make sense as a flag for us because we have different criteria vs the web request api